### ISSUE TITLE
Transpilation pain and compatibility

### ISSUE THREAD

**Issue by sithmel** [sithmel/iter-tools#9]
We need work around the way iter-tools is packaged, and of course discuss compatibility.
I think is reasonable having these class of environments supported:
* old browsers: IE11 (I know it is a pain but for many people lacking of IE11 support is a show stopper)
* current/green browsers (no async iter)
* future browsers
* node 8
* node 10
* future node

I can the issue better with a table:

|                  | ES2015* | Iterator | Async Iterator |
|------------------|---------|----------|----------------|
| Node 8           |    X    |     V    |        X       |
| Node 10          |   X**   |     V    |        V       |
| Future Node      |    V    |     V    |        V       |
| IE11             |    V    |     X    |        X       |
| Current browsers |    V    |     V    |        X       |
| Future Browsers  |    V    |     V    |        V       |

\* native or through webpack/rollup
\** need to use a flag so It is unlikely will be used on production that way

We can sum up like this:
* no transpilation: future node and browsers
* transpile only  ES2015: node 10
* transpile ES2015, async-iter: node 8
* transpile async-iter: current browsers
* transpile async-iter, and iter (ES5): IE11

:sob: 

What's your view? @conartist6

**Comment by sithmel** [comment]
I have realised there is also the case of a library that uses iter-tools as a dependency.
What should they use?

This is what I would expect:
browser and node future: untranspiled version
node 8/10: require version

But what about async iter compatibility and node. Is there any point in including async iterables where they are no supported by the environment (streams don't expose async iterables in node 8, for example).

double :sob:

**Comment by sithmel** [comment]
Here's a proposal:

Splitting the package in 2:
* iter-tools
* async-iter-tools

and only taking care of transpiling es2015 imports.

Notifying in the readme that they require transpilation where iterators or async iterators are not supported.

**Comment by conartist6** [comment]
First I love the idea that everyone gets exactly the code that they know how to run, but that doesn't exist yet. Right now the status quo is that package authors transpile down to es5 language features.

The only place that a split is actually starting to form is with .mjs files, where the benefits of not transpiling imports/exports are substantial for static analysis, and there is a reasonable mechanism for the support haves and have nots to choose the correct file based on whether they know how to parse it. Note though that in many cases the expectation is still that other than import/export the files will be using only es5 features. Presently (and still in my diff) these people become second-class citizens by being forced to use the commonjs files which cannot easily be statically analyzed, and are forced to use different paths (/es5). This group is almost certainly the most major constituency, so sacrifices like that seem ill advised.

The use case for this library is still a little unique though since it's meant for use with iterators and generators and there's a category of people who are leveraging these features in their non-library code particularly because they know that the support landscape for it meets their needs.

I think the split should be between, say, iter-tools, and iter-tools-es, where both versions export both .js and .mjs modules, but iter-tools compiles everything else down to es2015 and is recommended for use by library authors, and iter-tools-es transpiles only transpiles async iterators. When the support landscape fo async iterators becomes broad enough you can do a major version bump on iter-tools-es and drop the transpilation step. You can merge iter-tools-es back into iter-tools (and deprecate the -es package) as soon as anyone has any idea how to manage an ecosystem that has a target of anything other than es5 syntax.

**Comment by conartist6** [comment]
The good news is that if you want to publish two packages it's quite easy to do in my diff.

**Comment by sithmel** [comment]
What about this:
Let's keep a single package and single index.js containing sync and async functions. We transpile in these formats:

|transpile | es2015 modules | iterators | async iterators | for|
|------------|------------------------|-------------|---------------------|----|
|          |                |           |                 | future node/browsers|
|          |        V       |           |                 | node 10|
|          |        V       |           |        V        | node 8|
|          |                |     V     |        V        | IE11|
|          |                |     V     |                 | current browsers|

Also I would keep a different folder for every format.

@conartist6 I saw in your PR you have .mjs and .js in the same folder, that might be confusing for library consumers.

**Comment by sithmel** [comment]
Or even:

|transpile | es2015 modules | async iterators | iterators | for|
|------------|------------------------|-------------|---------------------|----|
|          |                |           |                 | future node/browsers|
|          |        V      |           |                 | node 10/current browsers with browserify|
|          |        V      |     V     |                   |node 8/current browsers with browserify|
|          |        V      |     V     |        V        |IE11 with browserify|
|          |                |     V     |                  |current browsers with webpack/rollup|
|          |                |     V     |        V        |IE11 with webpack/rollup|

**Comment by sithmel** [comment]
I prepared and experiment to check how does import work.

main package
------------------
package.json
```json
{
  "name": "testmodule",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "webpack": "^4.9.1",
    "webpack-cli": "^2.1.4"
  }
}
```
a.mjs
```js
import test from 'test1';
console.log(test);
```
b.js
```js
const test = require('test1');
console.log(test);
```
webpack.config.js
```js
const path = require('path');

module.exports = [
  {
    entry: './a.mjs',
    output: {
      filename: 'abundle.js',
    }
  },
  {
    entry: './b.js',
    output: {
      filename: 'bbundle.js',
    }
  },
];
```

package "test1"
--------------------
package.json
```json
{
  "name": "test1",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "module": "index.mjs",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}
```
index.js
```js
module.exports = 'commonjs';
```
index.mjs
```js
export const value = 'es2015';
```

test using node
============
I have tested the behaviour of running a.mjs and b.js (a.mjs require node --experimental-modules).
The result is that only the "main" field in package.json is taken into consideration. "module" is ignored.
The behaviour is probably fine giving we don't care about treeshaking while using node.

test using webpack
===============
Things gets really strange here:
"module" is definitely taken into consideration but:

* a.mjs import the commonjs module
* b.js import the es2015 module

**Comment by conartist6** [comment]
I think the esmodule transition pretty much is confusing. I've got maybe 4 hours into reading various opinions about it and nobody has yet convinced me that there exists any single set of best practices which if followed will eliminate the headaches.

I think publishing 5-6 copies of the code in the same package is significantly worse, and just passes all of the uncertainty straight along to users, who are likely to make incorrect decisions.

I think there are really only two main use cases:
1. People who need to support all browsers in wide circulation (library authors, commercial sites)
2. People who eyeball browser compatibility tables and and are willing to natively use what's supported by modern browsers (internal tools devs, next gen experiences, me)

For better or worse a fully transpiled library should work in every environment, so I think that's has to be the baseline way of supporting the first group until IE11's end of life (2025? ugh...).

The second group knows who they are; they will self select into it. That's why I think providing them the means to self-select (e.g. 'iter-tools-es' or 'iter-tools@esnext'?).

For my part I'm building an isomorphic es6 application. I'm doing it in part because I see no movement from the community on moving on from es5 as the standard compilation target, and I intentionally wanted to find the points of friction that were preventing that. Well, I guess I have.

**Comment by conartist6** [comment]
Interesting finding about node --experimental-modules not respecting package.json module field. I suspect that the module bundlers are all configurable in terms of what their order of preference is for file extensions.

**Comment by sithmel** [comment]
So you would go for:
es2015 module -> future browsers
commonjs -> current browsers/node
es5 (commonjs) -> everyone else

With the caveat that everyone that wants the best perf will provide the transpilation from the original e2015 source. is this correct ?

**Comment by conartist6** [comment]
I think it's just a full combination of two factors:

transpiled to es5 vs. transpiled to es6
and
es2015 module vs. commonjs module

I still think the module differences can be handled by publishing .js and .mjs side by side. That leaves only es5 vs. es6, which can safely be addressed without threat to isomorphic users by using a different package name or tag. Users for whom performance is a major concern will use the es6 version and run their code in an environment with es6 support.

By the way, webpack's default configuration for file extensions is: ["", ".webpack.js", ".web.js", ".js", ".json"]. However, create-react-app's default includes .mjs before .js.

**Comment by conartist6** [comment]
My reasoning is: if you do the simplest thing, it should just work. All other concerns, including tree shaking and perf should be secondary. You should be able to get incremental improvements if you know what you're doing, and I think this method allows you to:

If you want tree shaking, use create-react-app (next version which integrates webpack4, should be released soon) or configure your bundler to include .mjs before .js ones.

If you want better perf at the cost of IE11 support, use the -es or @next package.

**Comment by sithmel** [comment]
@conartist6 
I think I now understand your point of having .js and .mjs in the same folder.

But for a general adoption is important a better support of current environments. For example we can't treat node 8 as second class citizen.
And I am sure other users can be frustrated if their use case are getting a suboptimal version.

I am quite sold to have:
* untranspiled .mjs and .js the root (just like your PR)
* es5 compatible .mjs and .js in the es5 folder
* es2015 compatible .mjs and .js in the es2015 folder

In the long run es5 and es2015 folder can be dropped.

**Comment by conartist6** [comment]
Ok, great, I think that seems pretty much sensible in terms of what to offer and in what combinations. I hadn't counted in the fully untranspiled copies, but they'll be the smallest files so it won't add much weight.

I still think the root should be the es2015 (safe/universal) and then the folders can be for  semitranspiled (modern) and untranspiled (future).

**Comment by conartist6** [comment]
Give me a heads up before you push v4. I'll submit another PR with the changes we discussed here.

**Comment by sithmel** [comment]
@conartist6 I think the changes are all here:
https://github.com/sithmel/iter-tools/pull/10
 
I have also updated to babel7 (there was an issue with the transpilation of "asyncExecute").
And some other minor change.

**Comment by sithmel** [comment]
@conartist6 master is now up to date with the latest changes.
I also updated the doc. It will be published soon

**Comment by sithmel** [comment]
@conartist6  One last thing. I have tried to publish with https://github.com/sithmel/iter-tools/blob/master/package.json#L24:
```
    "dist": "npm run build && copyfiles -f package.json LICENSE README.md dist/ && cd dist && json -I -f package.json -e \"this.main='index.js';this.typings='index.d.ts';this.module='index.mjs'\" && npm publish",
```
The problem is that, running "npm publish", it runs prepack. Prepack runs "build" and this is not good.

Also I am bit confused on how this can work when using 'npm link'

**Comment by sithmel** [comment]
Eventually I decided to keep it simple.
```js
const takeWhile = require('iter-tools').takeWhile; // ES5 is default
import { takeWhile } from 'iter-tools'; // ES5 is default
 
const takeWhile = require('iter-tools/es5').takeWhile; // ES5
const takeWhile = require('iter-tools/es5/take-while'); // ES5
import { takeWhile } from 'iter-tools/es5'; // ES5
 
const takeWhile = require('iter-tools/es2015').takeWhile; // ES2015
const takeWhile = require('iter-tools/es2015/take-while'); // ES2015
import { takeWhile } from 'iter-tools/es2015'; // ES2015
 
const takeWhile = require('iter-tools/es2018').takeWhile; // ES2018
const takeWhile = require('iter-tools/es2018/take-while'); // ES2018
import { takeWhile } from 'iter-tools/es2018'; // ES2018
```
Also I found the trick to make node work with the correct version: just not specifying the extension in the "main" field in package.json.

Published as 4.0.1
